ngi_reports/utils/entities.py

```diff
@@ -0,0 +1,453 @@
+        self.run_params = {}
+        self.chemistry = {}
+        self.casava = None
+        self.seq_software = {}
+
+class Lane:
+    """Lane class
+    """
+    def __init__(self):
+        self.avg_qval = ''
+        self.cluster  = ''
+        self.id       = ''
+        self.phix     = ''
+
+    def set_lane_info(self, to_set, key, lane_info, reads, as_million=False):
+        """Set the average value of gives key from given lane info
+        :param str to_set: class parameter to be set
+        :param str key: key to be fetched
+        :param dict lane_info: a dictionary with required lane info
+        :param str reads: number of reads for keys to be fetched
+        """
+        try:
+            v = np.mean([float(lane_info.get('{} R{}'.format(key, str(r)))) for r in range(1,int(reads)+1)])
+            val = '{:.2f}'.format(round(v/1000000, 2)) if as_million else '{:.2f}'.format(round(v, 2))
+        except TypeError:
+            val = None
+
+        if to_set == 'cluster':
+            self.cluster = val
+        elif to_set == 'avg_qval':
+            self.avg_qval = val
+        elif to_set == 'fc_phix':
+            self.phix = val
+
+class AbortedSampleInfo:
+    """Aborted Sample info class
+    """
+    def __init__(self, user_id, status):
+        self.status = status
+        self.user_id = user_id
+
+class Project:
+    """Project class
+    """
+    def __init__(self):
+        self.aborted_samples = OrderedDict()
+        self.samples = OrderedDict()
+        self.flowcells = {}
+        self.accredited = { 'library_preparation': 'N/A',
+                            'data_processing': 'N/A',
+                            'sequencing': 'N/A',
+                            'data_analysis':'N/A'
+                            }
+        self.application = ''
+        self.best_practice = False
+        self.cluster = ''
+        self.contact = ''
+        self.dates = { 'order_received': None,
+                        'open_date':  None,
+                        'first_initial_qc_start_date': None,
+                        'contract_received': None,
+                        'samples_received': None,
+                        'queue_date': None,
+                        'all_samples_sequenced': None
+                    }
+        self.is_finished_lib = False
+        self.is_hiseqx = False
+        self.library_construction  = ''
+        self.missing_fc  = False
+        self.ngi_facility = ''
+        self.ngi_name = ''
+        self.not_as_million  = False
+        self.num_samples = 0
+        self.num_lanes = 0
+        self.ngi_id = ''
+        self.reference = { 'genome': None,
+                            'organism': None }
+        self.report_date = ''
+        self.sequencing_setup = ''
+        self.skip_fastq = False
+        self.user_ID = ''
+
+    def populate(self, log, organism_names, **kwargs):
+
+        project = kwargs.get('project', '')
+        if not project:
+            raise SystemExit()
```

Comment:
This causes a silent exit which might not even indicate an error. If this is a true error, raise an exception with a message instead.

Reply:
Done!